File: lib/utils/titled_form_pane.rb
# Copyright (C) 2013 patrick
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

#/*
# * To change this template, choose Tools | Templates
# * and open the template in the editor.
# */
#package viewpane;
#
#import java.util.*;
#import javafx.beans.property.DoubleProperty;
#import javafx.beans.property.SimpleDoubleProperty;
#import javafx.geometry.HPos;
#import javafx.geometry.Insets;
#import javafx.geometry.Orientation;
#import javafx.geometry.VPos;
#import javafx.scene.*;
#import javafx.scene.layout.*;
#import javafx.util.Pair;

class SD::Utils::TitledFormPane < Java::javafx.scene.layout.Pane
  include JRubyFX

	EXPAND_TAG = "titled-form-pane-expand"

  fxml_accessor :hgap, SimpleDoubleProperty
  fxml_accessor :maxRightWidth, SimpleDoubleProperty

  def self.setExpand(child, value)
		setConstraint(child, EXPAND_TAG, value)
  end

	def self.getExpand(child)
		true == getConstraint(child, EXPAND_TAG)
  end

  def getExpand(child)
    self.class.getExpand(child)
  end

	def computePrefWidth(d)
		layoutChildren?(false)
		return @lastWidth
  end

	def computePrefHeight(d)
		return d
  end

	def layoutChildren()
		layoutChildren?(true)
  end

	def layoutChildren?(actualLayout)
		managed = getManagedChildren()
		insets = getInsets()
		width = getWidth()
    height = getHeight()
		height = getPrefHeight() if height < 1.9

		top = snapSpace(insets.top)
    left = snapSpace(insets.left)
    bottom = snapSpace(insets.bottom)
    right = snapSpace(insets.right)
		height = [2, height - top - bottom].max

		itemHeight = []
		managed.each do |node|
			itemHeight << computeChildPrefAreaHeight(node, Insets::EMPTY, -1)
    end
    itemHeight = itemHeight.max || 12

    # now build up columns
    cols = []
		colLength = (height / itemHeight * 2).floor
		col = [[],[]]
		onLeft = true
		length = 0

		managed.each { |node|
			if (onLeft)
				if (getExpand(node))
          if (length + 3 >= colLength)
            # are we the last item? if so, bump to next row
            cols << col
            col = [[],[]]
            length = 0
          end
          length +=1
					onLeft = false
					col[1] << nil
        end
				length +=1
        col[0] << node
      else
				length+=1
				if (getExpand(node))
					puts("Warning: Unbalanced tree in TitledFormPane")
					length += 2
					onLeft = false
					col[1] << nil
					col[1] << nil
					col[0] << node
				else
					col[1] << node
        end
      end
			if (length + 1 >= colLength && !onLeft)
				cols << col
				col = [[],[]]
				length = 0
      end
			onLeft ^= true
    }
		cols << col if col[0].length > 0
		numCols = 2 * cols.length
		widths = []
    # computes ths widths of the columns
		cols.each do |pair|
      onLeft = true
			pair.each do |tmp|
				wmax = [0]
        tmp.each do |node|
          wmax << computeChildPrefAreaWidth(node, Insets::EMPTY, -1) if node
        end
        maxr = getMaxRightWidth || Float::MAX
        maxr = Float::MAX if onLeft or maxr < 1
				widths << [wmax.max, maxr].min
        onLeft = false
      end
    end
		#now we can set stuff
		j = 0
		prefixX = left
		cols.each do |pair|
      i = 0
			pair.each do |tmp|
				prefixY = top
				wmax = widths[j]
				tmp.each do |node|
					if (actualLayout && node)
						#TODO: check bounds?
						layoutInArea(node, prefixX, prefixY, getExpand(node) ? wmax + widths[j+1] : wmax, itemHeight, itemHeight, Insets::EMPTY, defaultHPos(i, node), VPos::CENTER)
          end
					#node.resizeRelocate(prefixX, prefixY, wmax, itemHeight)
					prefixY += itemHeight
        end
				prefixX += wmax + (getHgap() || 0)
				j+=1
        i+=1
      end
    end

    prefixX -= (getHgap()||0) if cols.length > 1
		@lastWidth = prefixX + right
  end

	def defaultHPos(i, node)
    eResult = GridPane.getHalignment(node) == nil ? HPos::LEFT : GridPane.getHalignment(node);
		if i == 0
			getExpand(node) ? eResult : HPos::RIGHT;
    else
      eResult
    end
  end

	def getContentBias()
		return Orientation::VERTICAL
  end

	# manually copied from JFX sources since they are package private stupidly
	def computeChildPrefAreaWidth( child,  margin,  height)

    top = margin != nil ? snapSpace(margin.top) : 0;
    bottom = margin != nil ? snapSpace(margin.bottom) : 0;
    left = margin != nil ? snapSpace(margin.left) : 0;
    right = margin != nil ? snapSpace(margin.right) : 0;
    alt = -1;
		if (child.getContentBias() == Orientation::VERTICAL)
      # width depends on height
      alt = snapSize(boundedSize(height != -1 ? height - top - bottom
          : child.prefHeight(-1), child.minHeight(-1), child.maxHeight(-1)));
    end
    return left + snapSize(boundedSize(child.prefWidth(alt), child.minWidth(alt), child.maxWidth(alt))) + right;
  end

  def computeChildPrefAreaHeight(child,margin, width)
    top = margin != nil ? snapSpace(margin.getTop()) : 0;
    bottom = margin != nil ? snapSpace(margin.getBottom()) : 0;
    left = margin != nil ? snapSpace(margin.getLeft()) : 0;
    right = margin != nil ? snapSpace(margin.getRight()) : 0;
    alt = -1;
    if (child.getContentBias() == Orientation::HORIZONTAL)
      # width depends on height
      alt = snapSize(boundedSize(width != -1 ? width - left - right
          : child.prefWidth(-1), child.minWidth(-1), child.maxWidth(-1)));
    end
    return top + snapSize(boundedSize(child.prefHeight(alt), child.minHeight(alt), child.maxHeight(alt))) + bottom;
  end

  def boundedSize( value,  min,  max)
    # if max < value, return max
    # if min > value, return min
    # if min > max, return min
    return [[value, min].max, [min, max].max].min
  end

  def self.setConstraint(paramNode, paramObject1, paramObject2)

    if (paramObject2 == nil)
      paramNode.getProperties().remove(paramObject1);
    else

      paramNode.getProperties().put(paramObject1, paramObject2);
    end
    if (paramNode.getParent() != nil)
      paramNode.getParent().requestLayout();
    end
  end

  def self.getConstraint(paramNode, paramObject)
    if (paramNode.hasProperties())
      return paramNode.getProperties().get(paramObject)
    end
  end
end

File: staging/sfxdesigner/src/edu/wpi/first/sfx/designer/ui/TitledFormPane.java
/*
 * Copyright (C) 2015 patrick
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package edu.wpi.first.sfx.designer.ui;

import java.util.ArrayList;
import java.util.List;
import javafx.beans.property.DoubleProperty;
import javafx.beans.property.SimpleDoubleProperty;
import javafx.geometry.HPos;
import javafx.geometry.Insets;
import javafx.geometry.Orientation;
import javafx.geometry.VPos;
import javafx.scene.Node;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.Pane;

/**
 *
 * @author patrick
 */
public class TitledFormPane extends Pane
{

	private static final String EXPAND_TAG = "jtitled-form-pane-expand";

	private SimpleDoubleProperty hgap_ = new SimpleDoubleProperty(0);

	public double getHgap()
	{
		return hgap_.doubleValue();
	}

	public void setHgap(double value)
	{
		hgap_.setValue(value);
	}

	public DoubleProperty hgapProperty()
	{
		return hgap_;
	}

	private SimpleDoubleProperty maxRightWidth_ = new SimpleDoubleProperty(Double.MAX_VALUE);

	public double getMaxRightWidth()
	{
		return maxRightWidth_.doubleValue();
	}

	public void setMaxRightWidth(double value)
	{
		maxRightWidth_.setValue(value);
	}

	public DoubleProperty maxRightWidthProperty()
	{
		return maxRightWidth_;
	}

	static public void setExpand(Node child, boolean value)
	{
		setConstraint(child, EXPAND_TAG, value);
	}

	static public boolean getExpand(Node child)
	{
		Boolean bol = (Boolean) getConstraint(child, EXPAND_TAG);
		if (bol == null)
			return false;
		return true == bol;
	}
	private double lastWidth;

	@Override
	protected double computePrefWidth(double d)
	{
		layoutChildren_(false);
		return this.lastWidth;
	}

	@Override
	protected double computePrefHeight(double d)
	{
		return d;
	}

	public void layoutChildren()
	{
		layoutChildren_(true);
	}

	private void layoutChildren_(boolean actualLayout)
	{
		List<Node> managed = getManagedChildren();
		Insets insets = getInsets();
		double width = getWidth();
		double height = getHeight();
		if (height < 1.9)
		{
			height = getPrefHeight();
		}

		double top = snapSpace(insets.getTop());
		double left = snapSpace(insets.getLeft());
		double bottom = snapSpace(insets.getBottom());
		double right = snapSpace(insets.getRight());
		height = Math.max(2, height - top - bottom);

		final ArrayList<Double> itemHeight = new ArrayList<>();
		managed.forEach((Node node) ->
		{
			itemHeight.add(computeChildPrefAreaHeight(node, Insets.EMPTY, -1));
		}
		);
		double itemHeightM = itemHeight.stream().max(Double::compare).orElse(12.0);

		// now build up columns
		final ArrayList<ArrayList<ArrayList<Node>>> cols = new ArrayList<>();
		double colLength = Math.floor(height / itemHeightM * 2.0);
		ArrayList<ArrayList<Node>> col = new ArrayList<>();
		col.add(new ArrayList<>());
		col.add(new ArrayList<>());
		boolean onLeft = true;
		int length = 0;

		for (Node node : managed)
		{
			if (onLeft)
			{
				if (getExpand(node))
				{
					if (length + 3 >= colLength)
					{
						// are we the last item? if so, bump to next row
						cols.add(col);
						col = new ArrayList<>();
						col.add(new ArrayList<>());
						col.add(new ArrayList<>());
						length = 0;
					}
					length += 1;
					onLeft = false;
					col.get(1).add(null);
				}
				length += 1;
				col.get(0).add(node);
			}
			else
			{
				length += 1;
				if (getExpand(node))
				{
					System.out.println("Warning: Unbalanced tree in TitledFormPane");
					length += 2;
					onLeft = false;
					col.get(1).add(null);
					col.get(1).add(null);
					col.get(0).add(node);
				}
				else
				{
					col.get(1).add(node);
				}
			}
			if (length + 1 >= colLength && !onLeft)
			{
				cols.add(col);
				col = new ArrayList<>();
				col.add(new ArrayList<>());
				col.add(new ArrayList<>());
				length = 0;
			}
			onLeft ^= true;
		}
		if (col.get(0).size() > 0)
		{
			cols.add(col);
		}

		int numCols = 2 * cols.size();
		ArrayList<Double> widths = new ArrayList<>();
		// computes ths widths of the columns
		for (ArrayList<ArrayList<Node>> pair : cols)
		{
			onLeft = true;
			for (ArrayList<Node> tmp : pair)
			{
				ArrayList<Double> wmax = new ArrayList<>();
				wmax.add(0.0);
				for (Node node : tmp)
				{
					if (node != null)
					{
						wmax.add(computeChildPrefAreaWidth(node, Insets.EMPTY, -1));
					}
				}
				double maxr = getMaxRightWidth();// || Double.MAX_VALUE;
				if (onLeft || maxr < 1)
				{
					maxr = Double.MAX_VALUE;
				}
				widths.add(Math.min(maxr, wmax.stream().max(Double::compare).get()));
				onLeft = false;
			}
		}
		//now we can set stuff
		int j = 0;
		double prefixX = left;
		for (ArrayList<ArrayList<Node>> pair : cols)
		{
			int i = 0;
			for (ArrayList<Node> tmp : pair)
			{
				double prefixY = top;
				double wmax = widths.get(j);
				for (Node node : tmp)
				{
					if (actualLayout && node != null)
					{
						//TODO: check bounds?
						layoutInArea(node, prefixX, prefixY,
									 getExpand(node) ? wmax + widths.get(j + 1) : wmax, itemHeightM, itemHeightM, Insets.EMPTY,
									 defaultHPos(i, node), VPos.CENTER);
					}
					//node.resizeRelocate(prefixX, prefixY, wmax, itemHeight)
					prefixY += itemHeightM;
				}
				prefixX += wmax + (getHgap());
				j += 1;
				i += 1;
			}
		}
		if (cols.size() > 1)
		{
			prefixX -= (getHgap());
		}
		this.lastWidth = prefixX + right;
	}

	private HPos defaultHPos(int i, Node node)
	{
		HPos eResult = GridPane.getHalignment(node) == null ? HPos.LEFT : GridPane.getHalignment(node);
		if (i == 0)
		{
			return getExpand(node) ? eResult : HPos.RIGHT;
		}
		else
		{
			return eResult;
		}

	}

	@Override
	public Orientation getContentBias()
	{
		return Orientation.VERTICAL;
	}

	// manually copied from JFX sources since they are package private stupidly
	private double computeChildPrefAreaWidth(Node child, Insets margin, double height)
	{

		double top = margin != null ? snapSpace(margin.getTop()) : 0;
		double bottom = margin != null ? snapSpace(margin.getBottom()) : 0;
		double left = margin != null ? snapSpace(margin.getLeft()) : 0;
		double right = margin != null ? snapSpace(margin.getRight()) : 0;
		double alt = -1;
		if (child.getContentBias() == Orientation.VERTICAL)
		{
			//# width depends on height
			alt = snapSize(boundedSize(height != -1 ? height - top - bottom
									   : child.prefHeight(-1), child.minHeight(-1), child.maxHeight(-1)));
		}
		double tmp = left + snapSize(boundedSize(child.prefWidth(alt), child.minWidth(alt), child.maxWidth(alt))) + right;
		return tmp;
	}

	private double computeChildPrefAreaHeight(Node child, Insets margin, double width)
	{

		double top = margin != null ? snapSpace(margin.getTop()) : 0;
		double bottom = margin != null ? snapSpace(margin.getBottom()) : 0;
		double left = margin != null ? snapSpace(margin.getLeft()) : 0;
		double right = margin != null ? snapSpace(margin.getRight()) : 0;
		double alt = -1;
		if (child.getContentBias() == Orientation.HORIZONTAL)
		{
			// width depends on height
			alt = snapSize(boundedSize(width != -1 ? width - left - right
									   : child.prefWidth(-1), child.minWidth(-1), child.maxWidth(-1)));
		}
		return top + snapSize(boundedSize(child.prefHeight(alt), child.minHeight(alt), child.maxHeight(alt))) + bottom;
	}

	private double boundedSize(double value, double min, double max)
	{
		return Math.min(Math.max(value, min), Math.max(min, max));
	}

	static private void setConstraint(Node paramNode, Object paramObject1, Object paramObject2)
	{
		if (paramObject2 == null)
		{
			paramNode.getProperties().remove(paramObject1);
		}
		else

		{
			paramNode.getProperties().put(paramObject1, paramObject2);
		}

		if (paramNode.getParent() != null)
		{
			paramNode.getParent().requestLayout();
		}

	}

	static Object getConstraint(Node paramNode, Object paramObject)
	{
		if (paramNode.hasProperties())
		{
			return paramNode.getProperties().get(paramObject);
		}
		return null;
	}
}
